### DISCUSSION TITLE
Using urql suspense feature with React 18

### DISCUSSION THREAD

**Discussion by jwm0** [urql-graphql/urql#2235] [Q&A]
Hi! It looks like urql has had `suspense` flag for a while but I'm having issues getting it to work with latest React 18 (RC). Am I doing something wrong or it's expected not to work on R18 yet? It looks like the suspense boundary never goes away (even though the network request completes).

Reproduction:
CRA with React 18
```json
    "react": "^18.0.0-rc.0",
    "react-dom": "^18.0.0-rc.0",
```

```jsx
// index.ts
const client = createClient({
  url: "...",
  suspense: true,
});

const root = ReactDOM.createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <Provider value={client}>
      <App />
    </Provider>
  </React.StrictMode>
);
```

```jsx
// App

const SuspenseTest = () => {
  const [{ data }] = useCheckoutQuery({ // hook auto-generated from @graphql/codegen
    variables: { ... },
  });

  return <span>{data.email}</span>;
};

export const App = () => {
  return (
      <p>
        <Suspense fallback={"Loading..."}>
          <SuspenseTest />
        </Suspense>
      </p>
  );
};

```

**Comment by JoviDeCroock** [comment]
Been able to reproduce it [here](https://codesandbox.io/s/stoic-pond-q81st?file=/src/App.js)

In React 18 we seem to be causing a loop [here](https://github.com/FormidableLabs/urql/blob/main/packages/react-urql/src/hooks/useQuery.ts#L122-L124) which could be caused by a promise overriding the `cache-entry`, will look into this later

**Reply by JoviDeCroock** [reply]
I've added a fix in https://github.com/FormidableLabs/urql/pull/2237

**Reply by JoviDeCroock** [reply]
This should be fixed in `2.1.3`, feel free to reopen if the issue persists, the linked PR's contain the reproduction used

**Reply by spikebrehm** [reply]
still getting this on `urql@2.2.3`

* `react@18.2.0`
* `next@13.4.6`

maybe fixed in a later version?

**Reply by JoviDeCroock** [reply]
@spikebrehm maybe there are more cases that didn't work in v2, we were very much in the exploration phase of `suspense` back then and have a more solid story nowadays. Do note that Next has a few more caveats with regards to i.e. potentially destroying the clients due to a missing `Suspense` boundary.

Looking at the changelog, we don't have that much more fixes after your version that involve `Suspense`, if possible mind checking this either in a sandbox against latest or explaining the issue? If it's sensitive feel free to do that in a DM

**Reply by spikebrehm** [reply]
Thanks @JoviDeCroock! I wonder if we should prioritize upgrading to 3.x or 4.x, assuming suspense support is more solid there?

**Reply by JoviDeCroock** [reply]
It's hard for me to tell you that in all honesty 😅 not entirely sure what the bug is you are running into... The changes from v3 to v4 are quite minimal https://github.com/urql-graphql/urql/issues/3114 and v2 to v3 are also quite small...

Larger migrations might be induced by the auth-exchange...

Is the same thing happening for you lot where the `Suspense` boundary never goes away, is this a one-off or happens everywhere?

**Reply by spikebrehm** [reply]
Just cut a branch to upgrade to v4. Simple enough that was, nice work!

I can see the page load and render fine and then error out with:

```
Error: Too many re-renders. React limits the number of renders to prevent an infinite loop.
```

Here's what I'm trying — doing it right? I also tried commenting out all our custom changes.

```ts
export default function Outer() {
  return (
    <Suspense fallback={<LoadingPage />}>
      <ComponentThatFetches />
    </Suspense>
  );
}

function ComponentThatFetches() {
  const [result, refetchQuery] = useSomeQuery({
    requestPolicy: 'network-only',
    context: {
      suspense: true,
    },
  });

  return (
    // ...
  );
}
```



our exchanges aren't doing anyth

**Reply by JoviDeCroock** [reply]
If you memoize context then it will work 😅 you can also define it outside your component.

Let me see if I can build a warning for that

**Reply by spikebrehm** [reply]
Oh my! That works! Thanks @JoviDeCroock!

I also just tried adding `suspense: true` to my top-level `createClient()` and that works as well. 👍🏼

**Reply by omniverse** [reply]
Yes @JoviDeCroock! I needed to override my client suspense setting for a specific query and memoizing `context` did the job. Thank you.

**Comment by adikari** [comment]
I am using `urql` with suspense in Next 13 and I am running into infinite loop problem.

**Reply by spikebrehm** [reply]
Do you have a top-level `<Suspense>` boundary?

Also check this:
https://github.com/urql-graphql/urql/discussions/2235#discussioncomment-7921102

**Reply by isener** [reply]
If you don't want suspense on Nextjs and only intend to make request on client-side (browser only) - construct context object memoised unless you use React 19.

```
  const context = useMemo(
    () => ({
      suspense: false,
    }),
    [],
  );
  ```
